### PR TITLE
PPCDEV-379 Added possibility to modify visualizations of VisitorLog widget/view and Visitors Real-time widget. Refactored getLastVisitsStart() action to render report instead of using an API.

### PR DESCRIPTION
**Pull request by sebastianpiskorski**
… to customize vizualization template. Refactored getLastVisitsStart() action to render report instead of using an API.

### CODE REVIEW

File: core/Plugin/Visualization.php [hunk 1]
```diff
@@ -207,9 +207,29 @@ protected function buildView()
...
+         *
+         * Value of $templateFile should keep to Piwik format '@<Plugin_Name>/<template_name>.twig'
+         *
+         * The values under templateVars are accesible directly in template for example value under
+         * $templateVars['example'] in template is available under {{ example }}
+         *
+         * @param array &$templateFile The original template file, the overriding one should extend the original one
+         * @param array &$templateFile Template variables, you can add new or modify the ones already present
+         */
+        Piwik::postEvent('Visualization.customizeTemplate', array(&$templateFile, &$templateVars));
```
Comment (tsteur) [inline core/Plugin/Visualization.php]: This would make our template file names API. Instead it would be better to use Visualization ID. Also it would make the template vars API which is not ideal. Can we instead post an instance of the Visualization? Otherwise we make any template variable an API and won't be able to make any changes anymore

File: plugins/Live/Model.php [hunk 2]
```diff
@@ -395,7 +395,7 @@ public function makeLogVisitsQueryString($idSite, $period, $date, $segment, $off
         $groupBy = false;
         $limit = $limit >= 1 ? (int)$limit : 0;
         $offset = $offset >= 1 ? (int)$offset : 0;
-        $orderBy = "idsite, visit_last_action_time " . $filterSortOrder;
+        $orderBy = "visit_last_action_time " . $filterSortOrder;
```
Comment (tsteur) [inline plugins/Live/Model.php]: I think this is unrelated to this change. Can you issue a different PR for this?

File: plugins/Live/templates/_dataTableViz_visitorLog.twig [hunk 3]
```diff
@@ -190,6 +190,7 @@ GPS (lat/long): {{ visitor.getColumn('latitude') }},{{ visitor.getColumn('longit
             {% endif %}
 
             <div class="col-md-{% if displayVisitorsInOwnColumn %}7{% else %}8{% endif %} column {% if visitor.getColumn('visitConverted') and not isWidget %}highlightField{% endif %}">
+                {% block beforeActionsInfo %}{% endblock %}
```
Comment (tsteur) [inline plugins/Live/templates/_dataTableViz_visitorLog.twig]: Here you can use an event instead. I think it's like eg `Live_beforeActionsInfo|postEvent` or similar

File: plugins/Live/templates/_dataTableViz_visitorLog.twig [hunk 4]
```diff
@@ -205,6 +206,7 @@ GPS (lat/long): {{ visitor.getColumn('latitude') }},{{ visitor.getColumn('longit
                         {% include "@Live/_actionsList.twig" with {'actionDetails': visitor.getColumn('actionDetails')} %}
                     </ol>
                 </div>
+                {% block afterActionsInfo %}{% endblock %}
```
Comment (tsteur) [inline plugins/Live/templates/_dataTableViz_visitorLog.twig]: Same here, better use an event if needed to add something

File: plugins/Live/templates/getLastVisitsStart.twig [hunk 5]
```diff
@@ -2,11 +2,13 @@
 {% set maxPagesDisplayedByVisitor=100 %}
 
 <ul id='visitsLive'>
-    {% for visitor in visitors %}
+    {% for visitor in dataTable.getRows() %}
+        {% set idSite = visitor.idSite %}
```
Comment (tsteur) [inline plugins/Live/templates/getLastVisitsStart.twig]: Is the idSite variable used somewhere? Also is there a difference between `visitors` and `datatable.getRows()`?

File: plugins/Live/templates/getLastVisitsStart.twig [hunk 6]
```diff
@@ -2,11 +2,13 @@
...
         <li id="{{ visitor.idVisit }}" class="visit">
             <div style="display:none;" class="idvisit">{{ visitor.idVisit }}</div>
             <div title="{{ visitor.actionDetails|length }} {{ 'General_Actions'|translate }}" class="datetime">
                 <span style="display:none;" class="serverTimestamp">{{ visitor.serverTimestamp|raw }}</span>
+                {% block beforeVisitInfo %}{% endblock %}
```
Comment (tsteur) [inline plugins/Live/templates/getLastVisitsStart.twig]: Here as well, better use an event if needed

File: plugins/Live/Reports/GetLastVisitsDetailsWidget.php [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+ * Unauthorized copying of this file, via any medium is strictly prohibited.
+ *
+ * @link http://piwik.pro
+ */
+
+namespace Piwik\Plugins\Live\Reports;
+
+use Piwik\Plugins\Live\Visualizations\VisitorLogWidget;
+
+class GetLastVisitsDetailsWidget extends Base
```
Comment (tsteur) [inline plugins/Live/Reports/GetLastVisitsDetailsWidget.php]: This would possibly add it to ReportMetadata and break APIs / Apps (eg Piwik Mobile app) which cannot handle such a custom format of a report. I haven't tested it but I can imagine it will break stuff. 

It might help to define the method

```php
    public function configureReportMetadata(&$availableReports, $infos)
    {
    }
```
I'd recommend to test it with mobile app . Also we need to check re other side effects.

File: tests/PHPUnit/Integration/WidgetsListTest.php [hunk 8]
```diff
@@ -37,7 +37,7 @@ public function testGet()
         // check if each category has the right number of widgets
         $numberOfWidgets = array(
             'VisitsSummary_VisitsSummary'  => 6,
-            'Live!'                        => 4,
+            'Live!'                        => 5,
```
Comment (tsteur) [inline tests/PHPUnit/Integration/WidgetsListTest.php]: Please also add the widget info to https://github.com/piwik/piwik/blob/2.16.0-rc1/plugins/Widgetize/tests/System/WidgetTest.php#L829 . This test should automatically notice it is missing but noticed it is not noticing it automatically

File: core/Plugin/Visualization.php (hunk 1 above)
Comment (tsteur) [inline core/Plugin/Visualization.php]: We should actually completely remove this part and instead use specific events in the template files.

File: plugins/Live/API.php [hunk 9]
```diff
@@ -146,6 +146,14 @@ public function getLastVisitsForVisitor($visitorId, $idSite, $filter_limit = 10,
     }
 
     /**
+     * Proxy for @see getLastVisitsDetails
+     */
+    public function getLastVisitsDetailsWidget($idSite, $period = false, $date = false, $segment = false, $countVisitorsToFetch = false, $minTimestamp = false, $flat = false, $doNotFetchActions = false)
+    {
+        return $this->getLastVisitsDetails($idSite, $period, $date, $segment, $countVisitorsToFetch, $minTimestamp, $flat, $doNotFetchActions);
+    }
```
Comment (tsteur) [inline plugins/Live/API.php]: Not sure but why is there a proxy needed? Can we not use existing API method? That should be possible... let me know if there's a problem or help needed in order to use old API

File: plugins/Live/Model.php (hunk 2 above)
Comment (tsteur) [inline plugins/Live/Model.php]: Actually, this should be definitely removed from here for now.

File: plugins/Live/Reports/GetLastVisitsDetailsWidget.php [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+{
+    protected function init()
+    {
+        parent::init();
+        $this->widgetTitle = 'Live_VisitorLogWidget';
```
Comment (tsteur) [inline plugins/Live/Reports/GetLastVisitsDetailsWidget.php]: There is no translation for this key commited / pushed

**Comment by tsteur** [comment]
I'm a bit confused by the description of the PR since it also adds a widget etc. and not only posts an event. What is the widget supposed to do?

File: plugins/Live/Reports/GetLastVisitsDetailsWidget.php [hunk 11]
```diff
@@ -0,0 +1,32 @@
+<?php
+/**
+ * Copyright (C) Piwik PRO - All rights reserved.
```
Comment (tsteur) [inline plugins/Live/Reports/GetLastVisitsDetailsWidget.php]: The header should be Piwik.org everywhere

File: plugins/Live/Reports/GetLastVisitsDetailsWidget.php (hunk 7 above)
Comment (tsteur) [inline plugins/Live/Reports/GetLastVisitsDetailsWidget.php]: Also can we find a better name for this one? What is the difference of this one to the existing real time widget?

File: core/Plugin/Visualization.php (hunk 1 above)
Comment (sebastianpiskorski) [inline core/Plugin/Visualization.php]: I will  do so, but wouldn't this be useful to change Visualization or at least extend it, inject additional vars?

File: plugins/Live/templates/getLastVisitsStart.twig (hunk 5 above)
Comment (sebastianpiskorski) [inline plugins/Live/templates/getLastVisitsStart.twig]: Yes this is because now it is  visualization and thus there is no `visitors` variable but `dataTable`.

File: plugins/Live/Reports/GetLastVisitsDetailsWidget.php (hunk 7 above)
Comment (sebastianpiskorski) [inline plugins/Live/Reports/GetLastVisitsDetailsWidget.php]: This one displays the Widget for Dashbord. Where the original displays visualization for VisitorLog view page.

File: tests/PHPUnit/Integration/WidgetsListTest.php (hunk 8 above)
Comment (sebastianpiskorski) [inline tests/PHPUnit/Integration/WidgetsListTest.php]: It is already there, under https://github.com/piwik/piwik/blob/2.16.0-rc1/plugins/Widgetize/tests/System/WidgetTest.php#L178

I've only changed its rendering method to Report instead of using API and injecting data to template directly.

File: plugins/Live/API.php (hunk 9 above)
Comment (sebastianpiskorski) [inline plugins/Live/API.php]: Because each report requires its own API action because of construction of this method: https://github.com/piwik/piwik/blob/2.16.0-rc1/core/Plugin/Report.php#L326

The earlier implementation was a kind of hack I guess.

**Comment by sebastianpiskorski** [comment]
This PR doesn't add any widget. It just changes existing one. The way this widget is rendered. With this PR it is rendered as a Report.

File: plugins/Live/templates/_dataTableViz_visitorLog.twig (hunk 3 above)
Comment (sebastianpiskorski) [inline plugins/Live/templates/_dataTableViz_visitorLog.twig]: I don't know if posting events in templates is a good idea. If you loose all advantages of twig templates.

File: core/Plugin/Visualization.php (hunk 1 above)
Comment (tsteur) [inline core/Plugin/Visualization.php]: The problem is, it would be come API and we wouldn't be able to change it anymore.

Also a problem is that we do not really have a way yet, to document existing template variables etc

File: plugins/Live/API.php (hunk 9 above)
Comment (tsteur) [inline plugins/Live/API.php]: You could possibly overwrite the `fetch()` method in the Report then this Magic won't need a new method anymore. Let me know how it goes, happy to help.

File: plugins/Live/Reports/GetLastVisitsDetailsWidget.php (hunk 7 above)
Comment (tsteur) [inline plugins/Live/Reports/GetLastVisitsDetailsWidget.php]: When I tried it, it showed my the real time widget and I couldn't notice a difference so I got confused as there's already a real time widget and a visitor log widget. Still not 100% sure what the difference is for that new one sorry.

File: plugins/Live/templates/_dataTableViz_visitorLog.twig (hunk 3 above)
Comment (tsteur) [inline plugins/Live/templates/_dataTableViz_visitorLog.twig]: We want to expose as less as possible as API. The events is a solution that will allow us easier to maintain it and to not break API. For this reason we cannot let anyone change view variables etc. Instead, for now, the solution is to use events unless we have possibly more stuff in Angular etc. 

The ideal solution would be to describe what you need to add and we could see if we can provide a stable API that let's any plugin do this. This way it would be completely unrelated to templates etc. But it's not always possible.

File: plugins/Live/templates/getLastVisitsStart.twig (hunk 5 above)
Comment (tsteur) [inline plugins/Live/templates/getLastVisitsStart.twig]: :+1:

File: plugins/Live/Reports/GetLastVisitsDetailsWidget.php (hunk 7 above)
Comment (tsteur) [inline plugins/Live/Reports/GetLastVisitsDetailsWidget.php]: OK I got it. It was changed to use Report / Visualization class etc so it can use the event I presume. It's not meant to be a new widget.

Comment (sebastianpiskorski) [inline plugins/Live/Reports/GetLastVisitsDetailsWidget.php]: I gave made a mistake and removed this report as a widget. I though that if it is defined  in other place it wont use this report setting. This plugin is a bit confusing.

File: tests/PHPUnit/System/expected/test_periodIsRange_dateIsLastN_MetadataAndNormalAPI__Live.getLastVisitsDetailsWidget_range.xml [hunk 12]
```diff
@@ -0,0 +1,412 @@
+<?xml version="1.0" encoding="utf-8" ?>
```
Comment (tsteur) [inline tests/PHPUnit/System/expected/test_periodIsRange_dateIsLastN_MetadataAndNormalAPI__Live.getLastVisitsDetailsWidget_range.xml]: I guess those files are not needed anymore?

**Comment by tsteur** [comment]
PR looks good. Need to wait for test results

File: tests/PHPUnit/System/expected/test_periodIsRange_dateIsLastN_MetadataAndNormalAPI__Live.getLastVisitsDetailsWidget_range.xml (hunk 12 above)
Comment (sebastianpiskorski) [inline tests/PHPUnit/System/expected/test_periodIsRange_dateIsLastN_MetadataAndNormalAPI__Live.getLastVisitsDetailsWidget_range.xml]: Not sure, will see at the test outcome.

Comment (sebastianpiskorski) [inline tests/PHPUnit/System/expected/test_periodIsRange_dateIsLastN_MetadataAndNormalAPI__Live.getLastVisitsDetailsWidget_range.xml]: You were right we don't need them any more.

**Comment by tsteur** [comment]
FYI: UI tests succeeded apart from 2 failing tests: http://builds-artifacts.piwik.org/piwik/piwik/master/17920 I don't know if it's a random build failure or due to the new visualization addition (I expect it is because of adding it). The tests may be fixed by adding

```php
    public static function canDisplayViewDataTable(ViewDataTable $view)
    {
        return ($view->requestConfig->getApiModuleToRequest() === 'Live');
    }
```
to the visualization

**Comment by sebastianpiskorski** [comment]
@tsteur I've added it, but test still fail.

**Comment by tsteur** [comment]
Quick question, in 2.16.0, would it be enough to only have the events in the templates and not the widget changes? I guess for this feature the visualization is not really needed for now and we could have it in 2.16.1?

**Comment by sebastianpiskorski** [comment]
I guess that would be fine too.

**Comment by tsteur** [comment]
Do you mind preparing quickly another PR that only contains the postEvents in the template? We would merge immediately and release 2.16.0. We can merge this one then for 2.16.1. We need to release 2.16.0 soon and problem is this change was not included in any beta yet. So we want to limit possible regressions.

**Comment by tsteur** [comment]
Let me know if you can't do it and I will issue the PR

**Comment by tsteur** [comment]
Done here https://github.com/piwik/piwik/pull/9693

**Comment by sebastianpiskorski** [comment]
I guess there is no point anymore as you have relased 2.16.0 already.

**Comment by mattab** [comment]
Closing as necessary changes should be in 2.16.0 (#9693)
